### ISSUE TITLE
Some suggestions about the type definition

### ISSUE THREAD

**Issue by fuafa** [rematch/rematch#661]
My bad, this is correct

~~## The order of overloads matters~~

~~We use a lot of intersection type of functions, which is the same as function overloads. According to TypeScript official docs, function overloads should from specific to general so that we can prevent from always resolving to general cases (any).~~

~~Example:~~

~~The order of the first 3 should be revert.~~

```ts
export type ExtractRematchDispatcherAsyncFromEffect<E> =
  E extends () => Promise<any> ? RematchDispatcherAsync<void, void>
  E extends (payload: infer P) => Promise<any> ? RematchDispatcherAsync<P, void> : 
  E extends (payload: infer P, meta: infer M) => Promise<any> ? RematchDispatcherAsync<P, void> :
  RematchDispatcherAsync<any, any> 
```

## The order of branches of conditional type matters

Same as the above.

Example:

`P extends any` always be true (except P = `never`), we will never reach the last two branches.

```ts
export type RematchDispatcherAsync<P = void, M = void> =
  ((action: Action<P, M>) => Promise<Redux.Dispatch<Action<P, M>>>) &
  ((action: Action<P, void>) => Promise<Redux.Dispatch<Action<P, void>>>) &
  (P extends void ? ((...args: any[]) => Promise<Action<any, any>>) :
    P extends any ? ((payload: any) => Promise<Action<any, any>>) :
    M extends void ? ((payload: P) => Promise<Action<P, void>>) :
    (payload: P, meta: M) => Promise<Action<P, M>>)
```

## Finally
Thank you for having such a coooool project.

**Comment by semoal** [comment]
There are already some open pull request that resolves this issue, possibly this week i take a look to all of them and we'll choose the best option.

Feel free to check them out if you want.

**Comment by ShMcK** [comment]
Should be resolved in @rematch/core v1.2 in #657.
